### PR TITLE
Drop last modified apis, update build script checker to use digests

### PR DESCRIPTION
**Pull request by jakemac53**
Lots of small changes to the test infrastructure as well here:

- `DatedValue`, `DatedString`, and `DatedBytes` are all gone.
- Added the `decodedMatches` matcher.
- Added ability to skip the build script check for testing purposes.
- Moved build script update tests to integration tests.

### CODE REVIEW

**Comment by natebosch** [comment]
Travis found some issues.

File: build_runner/lib/src/changes/build_script_updates.dart [hunk 1]
```diff
@@ -3,58 +3,140 @@
...
-import 'package:path/path.dart' as path;
+import 'package:path/path.dart' as p;
 
 import '../asset/reader.dart';
+import '../asset_graph/graph.dart';
 import '../generate/options.dart';
 
 class BuildScriptUpdates {
-  final RunnerAssetReader _reader;
+  final AssetGraph _graph;
```
Comment (natebosch) [inline build_runner/lib/src/changes/build_script_updates.dart]: can we get away with passing what is necessary as arguments to `hasBeenUpdated`?

File: build_runner/lib/src/changes/build_script_updates.dart [hunk 2]
```diff
@@ -3,58 +3,140 @@
...
-    var urisInUse = currentMirrorSystem().libraries.keys;
-    var updateTimes = await Future.wait(urisInUse.map(_uriUpdateTime));
-    return updateTimes.any((u) => u.isAfter(time));
+  Iterable<Uri> get _urisForThisScript => currentMirrorSystem().libraries.keys;
+
+  /// Checks if the current running program has been updated, based on previous
+  /// known digests in [_graph].
+  Future<bool> hasBeenUpdated() async {
+    return (await Future.wait(_urisForThisScript.map(_wasUpdated)))
+        .any((wasUpdated) => wasUpdated);
```
Comment (natebosch) [inline build_runner/lib/src/changes/build_script_updates.dart]: [nit] can this be `.contains(true)`?

File: build_runner/lib/src/changes/build_script_updates.dart [hunk 3]
```diff
@@ -3,58 +3,140 @@
...
+
+  /// Checks whether [uri] has been updated.
+  ///
+  /// If it can't tell due to an unsupported uri, then this always returns
+  /// `true`.
+  Future<bool> _wasUpdated(Uri uri) async {
+    AssetId id;
+    try {
+      id = _idForUri(uri);
+      // `null` but no error is a signal to skip this uri.
```
Comment (natebosch) [inline build_runner/lib/src/changes/build_script_updates.dart]: call this out in the Doc comment

File: build_runner/lib/src/changes/build_script_updates.dart [hunk 4]
```diff
@@ -3,58 +3,140 @@
...
+      return true;
+    }
+    if (node.lastKnownDigest == null) {
+      // This could happen if a new import was added, or we weren't able to
+      // initially compute the last known digest.
+      return true;
+    } else {
+      var currentDigest = await _reader.digest(id);
+      if (node.lastKnownDigest == currentDigest) return false;
+      return true;
```
Comment (natebosch) [inline build_runner/lib/src/changes/build_script_updates.dart]: [nit] Return a condition rather thatn `if (condition) return false;`

```dart
else {
  return node.lastKnownDigest != await _reader.digest(id);
}
```

File: build_runner/lib/src/changes/build_script_updates.dart (hunk 1 above)
Comment (jakemac53) [reply build_runner/lib/src/changes/build_script_updates.dart]: We could, but we would have to pass it through a few different functions - not sure what we would gain?

File: build_runner/lib/src/changes/build_script_updates.dart (hunk 2 above)
Comment (jakemac53) [reply build_runner/lib/src/changes/build_script_updates.dart]: yep! done

File: build_runner/lib/src/changes/build_script_updates.dart (hunk 3 above)
Comment (jakemac53) [reply build_runner/lib/src/changes/build_script_updates.dart]: done

File: build_runner/lib/src/changes/build_script_updates.dart (hunk 4 above)
Comment (jakemac53) [reply build_runner/lib/src/changes/build_script_updates.dart]: done

**Pull request merged**